src/clojupyter/kernel/handle_event/comm_msg.clj

```diff
@@ -76,113 +67,144 @@
+        [A S])
       (handle-comm-msg-unknown ctx S comm-id))))
 
 (defmethod handle-comm-msg msgs/COMM-MSG-UPDATE
   [_ S {:keys [req-message] :as ctx}]
   (assert req-message)
-  (log/debug "received COMM:UPDATE")
-  (let [{{:keys [comm_id] {:keys [method state]} :data} :content} req-message]
+  (log/debug "Received COMM:UPDATE")
+  (let [{{:keys [comm_id] {:keys [state] :as data} :data} :content buffers :buffers} req-message]
     (assert comm_id)
     (assert state)
-    (if-let [comm-atom (comm-global-state/comm-atom-get S comm_id)]
-      (let [A (action (side-effect #(ca/state-update! comm-atom state)
-                                   {:op :update-agent :comm-id comm_id :new-state state}))]
-        (return ctx A S))
+    (if-let [comm-atom (get S comm_id)]
+      (let [state (insert-binary-buffers data buffers)
+            state (insert-comm-atoms S state)
+            A (action (side-effect #(swap! (.-comm-state_ comm-atom) merge state)
+                                    {:op :update-agent :comm-id comm_id :new-state state}))]
+          (log/debug "COMM:UPDATE Computed action: " A " and state: " S)
+          (log/debug "COMM:UPDATE state: " state)
+          (log/debug "COMM:UPDATE comm-id: " comm_id)
+          [A nil])
+      (do (log/warn "Received COMM-UPDATE with unknown comm_id: " comm_id " and state: " state)
+          (handle-comm-msg-unknown ctx S comm_id)))))
+
+(defmethod handle-comm-msg msgs/COMM-MSG-CUSTOM
+  [_ S {:keys [req-message] :as ctx}]
+  (assert req-message)
+  (let [{{:keys [comm_id] {{event :event :as content} :content :keys [method]} :data} :content} req-message
+        buffers (msgs/message-buffers req-message)]
+    (assert comm_id)
+    (assert (= method msgs/COMM-MSG-CUSTOM))
+    (log/debug "Received COMM-CUSTOM -- comm_id: " comm_id " content: " content "buffers: " buffers)
+    (if-let [comm-atom (get S comm_id)]
+      (let [k (keyword (str "on-" event))
+            state @comm-atom
+            callback (get-in state [:callbacks k] (constantly nil))]
+        (if (fn? callback)
+          (let [A (action (side-effect #(callback comm-atom content buffers) {:op :callback :comm-id comm_id :content content}))]
+            [A nil])
+          ;; If callback attr is not a fn, we assume it's a collection of fns.
+          (let [call (fn [] (doseq [f callback]
+                              (f comm-atom content buffers)))
+                A (action (side-effect call {:op :callback :comm-id comm_id :content content}))]
+            [A nil])))
       (handle-comm-msg-unknown ctx S comm_id))))
 
 (defmethod calc* msgs/COMM-MSG
   [_ S {:keys [req-message] :as ctx}]
   (assert req-message)
   (let [method (msgs/message-comm-method req-message)]
     (assert method)
-    (handle-comm-msg method S ctx)))
+    (try (handle-comm-msg method S ctx)
+      (catch Exception ex (log/error "Error while handling comm-msg: " (.getMessage ex))))))
```

Comment:
This doesn't work because the handler doesn't do the actual work, but returns an action that can fail somewhere else in the stack.